### DISCUSSION TITLE
Associations with external constraints

### DISCUSSION THREAD

**Discussion by muellerj** [jeremyevans/sequel#2188] [General]
I have a DB schema, where `User`, `Package` and `Role` interact in the following way:

``` ruby
    create_table(:rel_users_roles_package) do
      primary_key :id
      foreign_key :user_id, :users, :null=>false, :key=>[:id]
      foreign_key :role_id, :roles, :null=>false, :key=>[:id]
      foreign_key :package_id, :packages, :null=>false, :key=>[:id]
    end
```

``` ruby
class User < Sequel::Model
  many_to_many :roles, join_table: :rel_users_roles_package
end

class Package < Sequel::Model
  many_to_many :users, join_table: :rel_users_roles_package
end

class Role < Sequel::Model
   many_to_many :users, join_table: :rel_users_roles_package
   many_to_many :packages, join_table: :rel_users_roles_package
end
```

What I would like to do, is write something like the following:

``` ruby
user.add_package_role(package, role)
```

The closest I could come up with, is the following:

``` ruby
class User< Sequel::Model
  def add_package_role(package, role)
    # TODO: Write this prettier?
    DB[:rel_users_roles_package].insert(user_id: id, role_id: role.id, package_id: package.id)
  end

  def remove_package_role(package, role)
    DB[:rel_users_roles_package].where(user_id: id, role_id: role.id, package_id: package.id).delete
  end

  def package_roles(package)
    roles_dataset.where(package_id: package.id)
  end
end
```

This feels a little clunky to reference `DB` inside the model. Does anyone know of a better way to handle this? I was looking at the `condition` into `many_to_many`, but couldn't quite make that work.

**Comment by jeremyevans** [comment]
You could have a UserPackageRole model for the table.  Then you could have a one to many association in User, Package, and Role to that model.  That might make certain things easier.  However, what you have doesn't look bad.

There is nothing wrong with referencing DB inside your model.  If you don't want to reference the constant, you could use db, which is the model's database.

**Comment by muellerj** [comment]
As always, thanks for your prompt reply!On 19. Jul 2024, at 16:18, Jeremy Evans ***@***.***> wrote:﻿
You could have a UserPackageRole model for the table.  Then you could have a one to many association in User, Package, and Role to that model.  That might make certain things easier.  However, what you have doesn't look bad.
There is nothing wrong with referencing DB inside your model.  If you don't want to reference the constant, you could use db, which is the model's database.
